### ISSUE TITLE
How to support EGL

### ISSUE THREAD

**Issue by TheNeikos** [waysome/waysome#606]
After a nice chat on #wayland on freenode.net I found out how to use egl stuff and how we will use it.

Basically it goes like this:

On startup we create a GBMBuffer large enough for the display. We scan that one out (send it to the graphics card). 

This can be done as seen here: http://virtuousgeek.org/blog/index.php/jbarnes

That is for the setup. Now we basically have a gbm render target where we can use EGL calls to!

Then, when a client connects we get the buffer through 

``` c
gltex = eglCreateImageKHR(disp, NULL, EGL_WAYLAND_BUFFER_WL, &wl_buffer, NULL);
```

This gives us an EGLImage!
Then we use

``` c
glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, gltex);
```

Then, we have a `normal` 2d texture we can then simple render into the gbm buffer and tadaaaaaaa, we have stuff on the screen (given a page flip~)

**Comment by neithernut** [comment]
Nice, so as soon as the rendering is up and running, we can also start accepting drm buffers.

**Comment by TheNeikos** [comment]
No, no, this is for EGLBuffers! The part with `eglCreateImage` is what we use to get an eglImage from a wl_buffer! And we then get the texture of /that/ and composite it on the screen :)

**Comment by neithernut** [comment]
Ok, but it _still_ provides only additional functionality once we got the initial EGL compositing up and running (using shared memory buffers for now).

**Comment by TheNeikos** [comment]
I am not certain that it doesn't do that already. I asked in the irc again, just to be sure.

**Comment by neithernut** [comment]
If I understood the bit of documentation I found right, this _only_ concerns opaque handles for data which already _is_ in VRAM. The shm buffer data is not.

**Comment by TheNeikos** [comment]
Okay, so we talked a bit, he said nope, you can't (thanks god for that, would have been weird).

But he also said that using lots of textures is fine really as you practically do not run out of textures.

**Comment by TheNeikos** [comment]
(The guy is one of the developers of weston, well, a contributor to be exact)

Yup, basically he confirmed what we suspected. The vram is already an atlassed texture map!

**Comment by TheNeikos** [comment]
Now, with all that information, any ideas for a roadmap? I want this to be done /before/ friday! So this leaves today and tomorrow!!

**Comment by neithernut** [comment]
Well, ...
- [x] I got one PR which should pretty mush ready for merge, as soon as the CI has the GLES2 dev files installed.
- [x] Once that is merged, I can rebase my texture area branch and adapt it to the new buffer format representation (there are 3 lines to be changed, maybe 4). I _could_ rip out the entire texture area stuff, so we _only_ have textures, but I'm not quite sure about it. You may comment on this in #602.
- [x] Another thing which would go into that PR is adding either two textures _or_ texture areas to `ws_surface` and rewriting the commit/attach functions so it transfers the buffer on "attach" and performs a flip on "commit".
- [ ] Once we have textures, we can finally implement compositing ~~using the infrastructure which is partly laid out in #533~~. I hope we can perform this task within the next 24h or so.
- [x] After the HW accelerated composition is up and running, we can try to introduce EGL buffers.

We _could_ do the last two tasks in parallel (at least partly), but given my lack of experience in this area, I would prefer doing both by pair-programming, or leaving the latter one in your hands completely. We could also instruct @matthiasbeyer to implement the EGL buffer handling.

To stay on schedule (or not to fall further behind, for that matter), it's important that GLES2 development packages are installed on the CI and that a decision is made whether or not to drop texture areas in favor for plain textures.

**Comment by TheNeikos** [comment]
I would prefer that we drop the texture areas. It does not make sense to me to support that part of the code.

**Comment by matthiasbeyer** [comment]
I support everything which gives us functionality which shows the user that we can do stuff.

So, if I shall implement stuff, I will do it, of course!

If we have functionality which _could_, in any way, be implemented later (feb, march, ...) I would prefer to do it then and now fully focus on stuff which gives us user experience. Remember: After christmas there are only 2 weeks until the final presentation of the project!

**Comment by xVin** [comment]
I agree with @matthiasbeyer

**Comment by neithernut** [comment]
I agree, too, but in what way is it relevant here?
